File: 144_Binary_Tree_Preorder_Traversal/stack.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    Stack<TreeNode> stk = new Stack<TreeNode>();
    List<Integer> result = new ArrayList<Integer>();
    public List<Integer> preorderTraversal(TreeNode root) {
        
        if(root == null) {
            return result;
        }
        pushAll(root);
        while (!stk.isEmpty()) {
            TreeNode node = stk.pop();
            if (node.right != null) {
                pushAll(node.right);
            }
        }
        return result;
    }
    void pushAll(TreeNode node) {
        for (; node != null; node = node.left) {
            stk.push(node);
            result.add(node.val);
        }
    }
    
}

File: 915-inorder-predecessor-in-bst/stack.cc
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */

class Solution {
public:
    /**
     * @param root: the given BST
     * @param p: the given node
     * @return: the in-order predecessor of the given node in the BST
     */
    TreeNode * inorderPredecessor(TreeNode * root, TreeNode * p) {
        // write your code here
        if (root == NULL || p == NULL) {
            return NULL;
        }
        stack<TreeNode*> stk;
        TreeNode* cur = root;
        TreeNode* last = NULL;
        while (cur != NULL || !stk.empty()) {
            while (cur != NULL) {
                stk.push(cur);
                cur = cur->left;
            }
            cur = stk.top();
            stk.pop();
            if (cur == p) {
                return last;
            } else {
                last = cur;
            }
            cur = cur->right;
        }
        return NULL;
    }
};
